### PR TITLE
Users/pankagar/azure rg

### PR DESCRIPTION
**Pull request by pankagar13**

### CODE REVIEW

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 1]
```diff
@@ -60,6 +60,32 @@ function Restart-MachineInProvider
...
+    Write-Verbose "Entering in azure-initializer" -Verbose
+
     Switch-AzureMode AzureResourceManager
-    Write-Verbose "Switched to AzureResourceManager" -Verbose
+    if($machineGroup.ProviderDataList.Count -gt 0)
+    {
+        $providerDataName = $machineGroup.ProviderDataList[0].Name
+        Write-Verbose "ProviderDataName : $providerDataName" -Verbose
+        $providerData = Get-ProviderData -ProviderDataName $providerDataName -Connection $connection
+        $subscriptionName = $providerData.Properties.GetProperty("SubscriptionName")
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Should we move all these constants into a separate file and use it from there?

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 2]
```diff
@@ -60,6 +60,32 @@ function Restart-MachineInProvider
...
+        Write-Verbose "ProviderDataName : $providerDataName" -Verbose
+        $providerData = Get-ProviderData -ProviderDataName $providerDataName -Connection $connection
+        $subscriptionName = $providerData.Properties.GetProperty("SubscriptionName")     
+        $username = $providerData.Properties.GetProperty("Username")
+        $password = $providerData.Properties.GetProperty("Password")
+        Write-Verbose "SubscriptionName : $subscriptionName" -Verbose
+        Write-Verbose "Username : $username" -Verbose
+        $securePassword = ConvertTo-SecureString $password -AsPlainText -Force
+        $psCredential = New-Object System.Management.Automation.PSCredential ($username, $securePassword)
+        $azureAccount = Add-AzureAccount -Credential $psCredential
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Please validate that azure ps is imported.

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 3]
```diff
@@ -60,6 +60,32 @@ function Restart-MachineInProvider
 
 function Initialize-AzureHelper
 {
+    Write-Verbose "Entering in azure-initializer" -Verbose
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Import-AzurePowerShellModule this is a commandlet in the internal dll. Use this just to be sure that the module is loaded

Comment (pankagar13) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: added.

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 (hunk 1 above)
Comment (pankagar13) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: these variables are associated with providerdatalist only that's why I put them here only

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 (hunk 2 above)
Comment (pankagar13) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: using Import-AzurePowerShellModule

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 4]
```diff
@@ -58,8 +58,69 @@ function Restart-MachineInProvider
...
+        throw "Windows Azure Powershell module (Azure.psd1) not found."
+    }
+
+    $azureModuleFolder
+}
+
+function Import-AzurePowerShellModule
+{
+    # Try this to ensure the module is actually loaded...
+    $folder = Get-AzureModuleLocation
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: No need of explicitly adding this function here. Reuse the one defined in Internal dll.

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 5]
```diff
@@ -58,8 +58,69 @@ function Restart-MachineInProvider
     Write-Verbose "Restarted machine $machineName on Azure provider" -Verbose
 }
 
+function Get-AzureModuleLocation
+{
+    #Locations are from Web Platform Installer
+    $azureModuleFolder = ""
+    $azureX86Location = "${env:ProgramFiles(x86)}\Microsoft SDKs\Azure\PowerShell\ServiceManagement\Azure\Azure.psd1"
+    $azureLocation = "${env:ProgramFiles}\Microsoft SDKs\Azure\PowerShell\ServiceManagement\Azure\Azure.psd1"
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Same here

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 (hunk 3 above)
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Write-Verbose "Initializing azure provider" -Verbose

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 6]
```diff
@@ -60,6 +60,35 @@ function Restart-MachineInProvider
...
+            throw "There was an error with the Azure credentials used for machine group deployment"
+        }
+        Select-AzureSubscription -SubscriptionName $subscriptionName
+    }
+    else
+    {
+        throw "No providerdata is specified in machine group"
+    }
+
+    Write-Verbose "Leaving azure-initializer" -Verbose
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Initialized Azure resource provider

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 7]
```diff
@@ -60,6 +60,35 @@ function Restart-MachineInProvider
...
+        $password = $providerData.Properties.GetProperty("Password")
+        Write-Verbose "SubscriptionName : $subscriptionName" -Verbose
+        Write-Verbose "Username : $username" -Verbose
+        $securePassword = ConvertTo-SecureString $password -AsPlainText -Force
+        $psCredential = New-Object System.Management.Automation.PSCredential ($username, $securePassword)
+        $azureAccount = Add-AzureAccount -Credential $psCredential
+        if(!$azureAccount)
+        {
+            throw "There was an error with the Azure credentials used for machine group deployment"
+        }
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Established connection to Azure resource provider

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 8]
```diff
@@ -60,6 +60,35 @@ function Restart-MachineInProvider
...
+        Select-AzureSubscription -SubscriptionName $subscriptionName
+    }
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Selected $subscriptionName as the current subscription

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 9]
```diff
@@ -60,6 +60,35 @@ function Restart-MachineInProvider
...
+
+    Import-AzurePowerShellModule
+
     Switch-AzureMode AzureResourceManager
-    Write-Verbose "Switched to AzureResourceManager" -Verbose
+
+    if($machineGroup.ProviderDataList.Count -gt 0)
+    {
+        $providerDataName = $machineGroup.ProviderDataList[0].Name
+        Write-Verbose "ProviderDataName : $providerDataName" -Verbose
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Getting providerData: $providerDataName

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 (hunk 2 above)
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Do you want to capture the error in an error variable and display the error here, while doing add-azureaccount

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 10]
```diff
@@ -60,6 +60,35 @@ function Restart-MachineInProvider
...
+        $providerData = Get-ProviderData -ProviderDataName $providerDataName -Connection $connection
+        $subscriptionName = $providerData.Properties.GetProperty("SubscriptionName")     
+        $username = $providerData.Properties.GetProperty("Username")
+        $password = $providerData.Properties.GetProperty("Password")
+        Write-Verbose "SubscriptionName : $subscriptionName" -Verbose
```
Comment (abhinitmodi) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: Nullity checks? for password, username, subscriptionName? Shouldn't we throw?

File: Tasks/PerformActionOnMachineGroup/AzureHelper.ps1 [hunk 11]
```diff
@@ -60,6 +60,43 @@ function Restart-MachineInProvider
...
 function Initialize-AzureHelper
 {
+    Write-Verbose "Initializing azure resource provider" -Verbose
+
+    Import-AzurePowerShellModule
+
     Switch-AzureMode AzureResourceManager
-    Write-Verbose "Switched to AzureResourceManager" -Verbose
+
+    if($machineGroup.ProviderDataList.Count -gt 0)
```
Comment (pankajkhanzode) [inline Tasks/PerformActionOnMachineGroup/AzureHelper.ps1]: We should get into this code only after checking we have Azure Provider and also have valid Provider Data. If there is no provider data - we should not even get into Initialize-AzureHelper.

**Pull request merged**